Rework Bugtool

tommyp1ckles opened the pull request:
## Background: tldr

Bugtool was developed adhoc to solve the problem of providing a standard way of collecting debugging data from Cilium agents.  As time went on, more functionality has been added to accommodate more use cases. As well, with the creation of cilium-cli sysdump - the k8s exec functionality of bugtool is no longer necessary.  As a result, bugtools logic has become quite sprawling.

This set of changes seeks to do a major refactor of bugtool, with the following goals:

* Begin moving towards more structured/parsable output. Historically commands have been dumped primarily in human readable format (with some data being inconsistently dumped into JSON). We've come to a point where we need to start relying on tooling more for analyzing our bugtool data which is made very difficult by existing bugtool formatting. The goal being to provide bugtool data that can be parsed by tooling/scripts and avoiding redundancy.
    * **Note**: JSON formatted data should still be easy to read by a human (with maybe the help of jq).  There will be further PRs to address difficult to understand json formatted data such as `bpftool map dump ...`.
* Organize data into separate "topics" which can be toggled (i.e. providing a more granular way of doing dumps).
* Provide a clean and standard way of writing/scheduling dumping tasks, that allows for easier future maintenance and control of how they are executed.

### Follow up PRs
* Improve formatting of bpf map dumps, such that we can use Ciliums internal data struct definitions of bpf map key/values (for [example](https://github.com/cilium/cilium/blob/master/pkg/maps/ipcache/ipcache.go#L33))
    * This includes implementing json.Marshaller for various map key/value struct member types (i.e. types.IPv4).
* For versions v1.12+ use Hive for CLI entrypoint.

Fixes: #issue-number

```release-note
bugtool: rewrite much of bugtool code and move towards structured output. Also introduces new bugtool configuration format as well as --topics= flag for more granular dump specification. 
```

tommyp1ckles committed:
bugtool/options: add options for bugtool.

Options covers all options previously used by bugtool, with the exception
of removed functionality (i.e. k8s).

Most importantly, options are preserved such that existing builds of cilium-cli
are compatible with running "cilium sysdump" on agents with the refactored bugtool.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool: add wait option.

Bugtool dumps that happen prior to Agent being ready will be missing data.
This provides the option to wait until Cilium is ready before proceeding.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool: replace bugtool runTool code path with new bugtool code.

Replaces bugtool dump code path with new runTool that runs rebuilt
version of bugtool.

Also, removes all unused code.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool: improve disclaimer message with outline.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool: remove old bugtool code.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool/utils: move ipsec hashing functions to separate package.

This is to prevent import cycles in subsequent commits.
This preserves all logic and tests from original implementation.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool/cmd: remove cmd docs for removed k8s functionality.

Functionality was not kept in rewritten bugtool code paths.
K8s exec functionality is now performed by cilium-cli rendering
this code redundant and unmaintained.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool/configuration: Add package for building bugtool task tree.

Configuration builds a root dump.Task, that when run will perform a standard bugtool dump.

This is split into a few categories (i.e. "topics") that are implemented as dump.Dir's within the task.

* Agent: Dumps cilium agent cli debug data.
* BPFMaps: Dumps pinned cilium bpf maps and attachments/programs.
* System: Dumps relevant host system data.
* Envoy: Runs cilium-envoy specific dumps.
* State: Dumps cilium run state data (i.e. /run/cilium/...).
* PProf: Performs pprof profiling.
* Gops: Performs gops dumps.

These can be toggled via the cilium-bugtool flags. CreateDump constructs such a task, based on passed configuration.

The dump task provided seeks to provide all data that previous bugtool iteration had.
Except for the removal of duplicated commands that would dump both json and human readable data.

Encorporates work done in (addition of Envoy metrics task): 0307add (by Tam Mach <tam.mach@cilium.io>)

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool/dump: provide base for all task types.

Base will provide a compositional base for all bugtool dump tasks to be implemented.
This provides a single way to define a tasks type and kind, and to get an identifier for that task.

This also will provide a standard base to encode/decode tasks into config file format.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool/dump: add Task interface type.

This describes the behavior of a bugtool task that can be scheduled and run to dump debug data.

A task takes a dump.Context struct, which also provides the scheduling function.
This also handles collecting result data for the dump after executing task(s).

The premise is to provide a single way to write bugtool tasks that:

* Can reuse code for common dump task types (ex. exec, http req, file copy etc).
* Can be scheduled, run and reported on (into results.yaml)
* Encodes/decodes from configuration files.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool/dump: add file writer that deletes empty files after close.

This is to help prevent unnecessary empty files when writing a dump.
Status'/errors will be written to the results.yaml alongside the tasks.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool/dump: add decoder which can decode yaml to a bugtool conf.

Using this we can encode/decode all tasks into a configuration format.
This configuration preserves directory structure, and can also define all
task types (previously - only "exec" type tasks where written/decoded from
the bugtool configuration).

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool/dump: implement various task types.

Implementing the dump.Task type to cover all existing use cases
for bugtool:

* Dir: Sub-directory used for grouping other tasks together and reporting results.
* Exec: Executes command or script and writes output to file.
* File: Copies file(s) from system into dump.
* Request: Makes HTTP request on either network or unix sockets.

The Dir task will also write out a structured results file containing
run status about a task, as well as relevant stats (i.e. memory/cpu usage).

These implementations will be use to construct a new bugtool configuration that
performs all dump tasks that currently exist in bugtool.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool/config: reimplements improve tc filter dump commands.

Reproduces fix to bugtool tc tasks by specifying filter direction.

Based on changes in 60eb3e6 (work by: Joe Stringer <joe@cilium.io>)

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool/configuration: add cilium cgroup list command.

Replicates functionality of: 2692df3 (by Alex Katsman <alexkats@google.com>)

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool/configuration: remove old svc lb map dumps.

Based on: 9fcb306

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

squeed commented on bugtool/dump/task.go:
```diff
@@ -0,0 +1,172 @@
+
+	// Identifier returns an identifier string of the task, including
+	// the type and name.
+	Identifier() string
```
What is the expected contract for an identifier? Must it be unique?

tommyp1ckles replied:
Yeah, it should be unique. Although there's nothing enforcing that - you might just run into wrong behavior...

tommyp1ckles replied:
Perhaps this should just be up to the internal implementation, it's not actually necessary to expose this in the interface.

tommyp1ckles replied:
Ok, thought about this a bit more, the external exposed Identifier is just for convenience. Internally tasks have to be unique, but ill add this to the validation step.

squeed commented on bugtool/dump/dir.go:
```diff
@@ -0,0 +1,58 @@
+		return fmt.Errorf("invalid dir %q: %w", d, err)
+	}
+	for _, t := range d.Tasks {
+		if err := t.Validate(ctx); err != nil {
```
Would it make sense to return a fancy new go multierr here? (Not necessary, just curious)

squeed commented on bugtool/dump/file.go:
```diff
@@ -0,0 +1,83 @@
+// File implements a Task that will copy a file into a dump.
+type File struct {
+	base               `mapstructure:",squash"`
+	Src                string
```
Would it make sense if this were a glob?

Can this be a directory?

squeed commented on bugtool/dump/file.go:
```diff
@@ -0,0 +1,83 @@
+type File struct {
+	base               `mapstructure:",squash"`
+	Src                string
+	ExcludeObjectFiles string `mapstructure:"exclude_object_files"`
```
Likewise, rather than a specific feature for ExcludeObjectFiles, what about a negative-match regex or glob?

tommyp1ckles replied:
Yeah good point, I was kind of just covering recently added functionality but having one off stuff be part of these feels wrong.

A think a general glob or something makes sense,.

squeed commented on bugtool/dump/request.go:
```diff
@@ -0,0 +1,131 @@
+
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("bad status: %s", resp.Status)
```
Probably should write the output anyways? Just in the case it's relevant?

squeed commented on bugtool/dump/request.go:
```diff
@@ -0,0 +1,131 @@
+		"file": file,
+	})
+
+	out, err := os.Create(file)
```
This should probably be your fancy remove-on-empty file.

squeed commented on bugtool/dump/request.go:
```diff
@@ -0,0 +1,131 @@
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("bad status: %s", resp.Status)
+	}
+	_, err = io.Copy(out, resp.Body)
```
I think it might be useful for the output to look a bit more like `curl -v` -- show me the request and response headers.

tommyp1ckles replied:
Hmm, I agree.  I'd like to keep the files as close to what they where before,  I think maybe having Request write two files would make sense.

Let me get back to you on this.

squeed commented on bugtool/options/options.go:
```diff
@@ -0,0 +1,100 @@
+	)
+	flags.IntVar(&bugtoolConf.PProfTraceSeconds, "pprof-trace-seconds", 180, "Amount of seconds used for pprof CPU traces")
+	flags.BoolVar(&bugtoolConf.Generate, "generate", false, "Create configuration file of all commands that would have been executed")
+	flags.BoolVar(&bugtoolConf.DryRun, "dry-run", false, "[DEPRECATED: use \"--generate\" instead] Create configuration file of all commands that would have been executed")
```
I think you can mark this flag as deprecated in the `flags` package.

squeed commented on bugtool/configuration/configuration.go:
```diff
@@ -0,0 +1,369 @@
+	ts := dump.Tasks{}
+	conf.Topics = deduplicate(conf.Topics)
+	log.Debugf("Final topics list %v", conf.Topics)
+	logs, unstructured, structured := defaultResources()
```
nit: should this be in the "system" case?

squeed commented on bugtool/options/options.go:
```diff
@@ -37,7 +37,8 @@ type Config struct {
 	Timeout            time.Duration `mapstructure:"timeout"`
-	Debug              bool
+	Debug              bool          `mapstructure:"debug"`
+	Wait               bool          `mapstructure:"wait"`
```
Does this need a separate timeout? Or is the general timeout sufficient?

tommyp1ckles replied:
Good idea 👍

squeed commented on bugtool/cmd/bugtool.go:
```diff
@@ -0,0 +1,187 @@
+	}
+}
+
+func waitForAgentReady(ctx context.Context) {
```
this function is awkward: if it times out, it just exits. It probably should return an error instead.

squeed reviewed (commented):
Looks really nice - a lot cleaner than before. A few comments, and unfortunately you've tickled some linter errors in unrelated code.

nebril commented on bugtool/configuration/bpftool.go:
```diff
@@ -0,0 +1,109 @@
+	var mountpoint string
+	if bpffsMountpoint, err := bpffsMountpoint(); err == nil {
+		mountpoint = bpffsMountpoint
+		ts = append(ts, mapDumpPinned(mountpoint,
```
Probably out of scope of this PR, but should we list all BPF maps and dump ones that start with `cilium` to future-proof this? Not sure if it's a good idea.

tommyp1ckles replied:
Hmm that might be a good idea.

I have a follow up PR where I introduce a new Task type `BPFMap[KT, KV]`.  Maybe with that we could do some kind of pattern matching for pinned maps.

nebril commented on bugtool/dump/decode.go:
```diff
@@ -0,0 +1,78 @@
+}
+
+// decodeMap decodes the map as a Base struct, in order to identify
+// the kind of task to decode, then it performs a full decod
```
nit: s/decod/decode

nebril commented on bugtool/configuration/configuration.go:
```diff
@@ -0,0 +1,371 @@
+	conf.Topics = deduplicate(conf.Topics)
+	log.Debugf("Final topics list %v", conf.Topics)
+	logs, unstructured, structured := defaultResources()
+	for _, topic := range conf.Topics {
```
Could we drop the `topics` list and just have a config yaml defined somewhere in this package that has default tasks? Or do we do anything programmatically that isn't possible via yaml configuration?

This yaml could be helpful as documentation for how to use the config files, and we could drop all the code that goes over topics since we already have whole `Decode` package that could generate tasks from hardcoded yaml.

tommyp1ckles replied:
> Or do we do anything programmatically that isn't possible via yaml configuration?

No, the topics is just a convenient shorthand for just running a logical subset of tasks.

My reasoning for it is:
* Its a clean way to replace the --get-pprof and --envoy-dump flags. i.e. --topics=envoy.
* It'll be easy/clean for tools consuming this to just specify a topic, instead of having to generate/copy config around.

nebril reviewed (approved):
Looks good overall! Some nits and one bigger question left inline, but approving the PR since my suggestions might go out of scope of this PR.

tommyp1ckles committed:
bugtool/dump: Make file exclude more general.

This is just a general glob pattern that is excluded from the file(s).
We can remove references to obj files.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool/options: deprecate flags with plfags MarkDeprecated.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool: add v3 backends maps to bugtool dump config.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
documentation: generate cmdref docs.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool: add cmdref to new bugtool root command.

This allows for printing bugtool command documentation as well
as generating documentation.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool: make parallel workers flag deterministic.

Cmdref checks depend on deterministic command reference generation.
Using the CPU count as the default may cause this to fail.
Instead move this logic to the runTool function.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

pchaigno reviewed (changes requested):
I've tested this on a GKE cluster and have noticed a few bugs, have some questions and some concerns.

### Questions
- I'm seeing a bunch of `results.yaml` files. What are those?
- When a command fails, rather than an empty file, could we keep only the .err file? I'm a bit worried we'll miss errors here and assume the output was just empty because we used to report the errors in the output file. Maybe it's just a matter of deleting empty files?

### Minor
- Some of the JSON output were not tested. For example `ip --json x s` and `ip --json x p` don't actually return JSON. Even if the flag was correct, it gets ignored anyway.
- Some files can have a broken JSON output. For example, on an IPv4-only cluster, `ip6-route-tables.json` contains just `[`.
- The `state` directory is itself in a `state` directory.

### Major
- Almost everything is now JSON. I'm all for enabling us to perform post-processing, but this is making things harder to inspect for humans. I'm worried about the impact it will have on our ability to investigate incidents. How will this be addressed? Can we block the v1.14.0 release on adressing this and revert this PR if we don't manage to invest cycles to fix that in time for v1.14.0?

tommyp1ckles committed:
bugtool: use less generic name for results files and add header

New name makes it easier to distinguish the purpose of the file.

As well, write header comment to yaml file describing the purpose of the file.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool/task: remove empty output files.

Similar to empty err files, if an output file has no data written then
the file will be excluded from the dump files.
This is to avoid unnecessary files when a command fails and only prints an error.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bptool/dump: Include full command into result status.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool/configuration: fix ip commands.

"ip" commands where incorrectly being prefixed with an extra "ip" causing them to fail.
This corrects the behavior and also fixes the tc queue command.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool/configuration: fix bpftool net/prog tasks.

Each command bpftool ... show type command was using the same task name
which was causing collisions.

This fixes that such that the names are based on the resource being listed.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool: move ip xfrm <state/policy> command tasks to unstructured.

The ip xfrm commands don't actually use the json flag.
Move the commands to unstructured and remove the unused json flag.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool: fix ip route output for ipv6.

Previously, the ip version was not specified when gathering tables, thus
this was causing confusing output when attempting to gather ip6 routes for
ip4 tables.

This specifies the routes ip version for both iterating tables, and dumping them.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool: copy only file contents of cilium runtime state dir.

This avoids having nested state/state directory, root level state
dir is kept as a Dir task to maintain consistency.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

pchaigno commented on bugtool/configuration/cilium.go:
```diff
@@ -0,0 +1,84 @@
+		"cilium metrics list",
+		"cilium fqdn cache list",
+		"cilium config -a",
+		"cilium encrypt status",
```
This one doesn't seem to support a proper JSON output; it outputs text.

tommyp1ckles replied:
Moved it to ".md" file for now.  I'll add it to the list of commands to fix.

pchaigno commented on bugtool/configuration/cilium.go:
```diff
@@ -0,0 +1,84 @@
+		"cilium policy selectors",
+		"cilium node list",
+		"cilium lrp list",
+		"cilium cgroup list",
```
This command doesn't exist.

tommyp1ckles replied:
fixed `cgroup` -> `cgroups`, json output is also broken. I'll add it to the list of commands to fix.

pchaigno commented on bugtool/configuration/cilium.go:
```diff
@@ -0,0 +1,84 @@
+		"cilium status",
+		"cilium identity list",
+		"cilium-health status",
+		"cilium policy get",
```
This one doesn't output proper JSON. Only part of the output is JSON.

tommyp1ckles replied:
looks like a bug in the command code, ill file a separate issue.

tommyp1ckles replied:
Weirdly, if you don't specify -o json it actually prints the valid json and then prints "Revision: ...".

tommyp1ckles replied:
Oh I see, it actually does do valid json with the flag, but the json object just becomes a flat string type after the first set of keys:

I think this is a similar issue to commands like `cilium bpf lb list`. 

We'll have both types for now - I'll fix this all when I address that.

pchaigno commented on bugtool/configuration/cilium.go:
```diff
@@ -0,0 +1,84 @@
+		"cilium endpoint list",
+		"cilium bpf bandwidth list",
+		"cilium bpf tunnel list",
+		"cilium bpf lb list",
```
The JSON output for this one seems weirdly broken.

tommyp1ckles replied:
Its not very well crafted json but it does print valid json for me:

ex.
```
{
  "10.96.0.10:53": [
    "10.244.0.186:53 (2) (1)",
    "10.244.0.90:53 (2) (2)",
    "0.0.0.0:0 (2) (0) [ClusterIP, non-routable]"
  ],
  "10.96.0.10:9153": [
    "10.244.0.186:9153 (3) (1)",
    "0.0.0.0:0 (3) (0) [ClusterIP, non-routable]",
    "10.244.0.90:9153 (3) (2)"
  ],
  "10.96.0.1:443": [
    "0.0.0.0:0 (1) (0) [ClusterIP, non-routable]",
    "172.19.0.3:6443 (1) (1)"
  ]
}
```

tommyp1ckles replied:
many of these use a generic dump-to-json function for these. I am planning to improve these in a separate PR.

tommyp1ckles replied:
For now, we'll have the human readable format as well.

pchaigno commented on bugtool/configuration/cilium.go:
```diff
@@ -0,0 +1,84 @@
+		"cilium endpoint list",
+		"cilium bpf bandwidth list",
+		"cilium bpf tunnel list",
+		"cilium bpf lb list",
```
The JSON output for this one seems weirdly broken.

tommyp1ckles replied:
Same as above

pchaigno commented on bugtool/configuration/configuration.go:
```diff
@@ -0,0 +1,379 @@
+			// system contains system level dumps.
+			ts = append(ts, dump.NewDir("system", dump.Tasks{
+				dump.NewDir("structured", structured),
+				dump.NewDir("unstructured", unstructured),
```
Why separate structured and unstructured command outputs? As a user, I typically go into the bugtool report looking for some specific information. I have an idea of which command will give me that information, but I won't remember if its output is among the structured or unstructured ones :-/

tommyp1ckles replied:
Yeah I think you're right, I combined them into a single directory: `cmd`.

pchaigno commented on bugtool/configuration/configuration.go:
```diff
@@ -0,0 +1,379 @@
+		"-j -6 r",
+		"-j -d -s l",
+		"-j -4 n",
+		"-j -6 n",
```
I'm not finding the output for these two but I have a `ip-n.json` file with a broken JSON content:
```json
[]
dst":"10.164.0.1","dev":"eth0","lladdr":"42:01:0a:a4:00:01","state":["REACHABLE"]},{"dst":"10.68.1.24","dev":"lxc4eac24edaf5e","lladdr":"a6:0a:28:ec:2b:ad","state":["REACHABLE"]},{"dst":"10.68.1.49","dev":"lxc_health","lladdr":"ee:b0:b4:e3:84:67","state":["DELAY"]},{"dst":"10.68.1.249","dev":"lxcaa96ecd074ee","lladdr":"ba:4a:5e:44:da:24","state":["REACHABLE"]},{"dst":"10.68.1.133","dev":"lxcb94237c3f348","lladdr":"de:a4:de:da:28:d4","state":["REACHABLE"]},{"dst":"10.68.1.193","dev":"lxcad1b39b5fa79","lladdr":"f6:ed:6b:7d:a1:04","state":["REACHABLE"]}]

```

tommyp1ckles replied:
Thats odd, could be a potential bug in the ip command. The json encoding it uses seems potentially error prone (was taking a look at why xfrm json output doesn't work).

tommyp1ckles replied:
Although I'm surprised it's different, considering the tool is part of the container image. I have no problems with it:

```
[{"dst":"fe80::149f:d2ff:fec5:5a74","dev":"lxc_health","lladdr":"16:9f:d2:c5:5a:74","state":["STALE"]},{"dst":"fc00:c111::3","dev":"eth0","lladdr":"02:42:ac:13:00:03","router":null,"state":["REACHABLE"]}]}]
```

Did you build the image using the dev image targets?

pchaigno commented on bugtool/configuration/bpftool.go:
```diff
@@ -0,0 +1,112 @@
+	ts := dump.Tasks{
+		newBPFMapTask("map", "show"),
+		newBPFMapTask("prog", "show"),
+		newBPFMapTask("net", "show"),
```
These last two are in the `bpfmaps` directory but they are not about BPF maps.

tommyp1ckles replied:
Switched it to be: `bpfmaps` -> `bpf` with a map sub-directory

pchaigno reviewed (changes requested):
(Uh, weird. GitHub posted two partial reviews above even though I'm sure I didn't hit the "Submit review" button before :-/)

I found a few other issues, including several broken JSON outputs :point_down: :point_up: 

> @pchaigno Fixed the minor issues, going to add the "parity" flag in a follow up PR.

Do you have an idea when that would be ready? I was focused on incident response before, but this change to all JSON is also going to impact our ability to debug CI flakes even if it's not in a release yet :grimacing:

tommyp1ckles committed:
bugtool: add human readable flag.

This will also dump various tasks without structured formatting.
While appropriate tooling is developed to help with fully structured
dumps, this will provide a transition where human readable analysis of
various dumps is still easy.

Also change naming of directories/topics to make navigation easier:
* bpfmaps -> bpf + bpf/maps)
* merged system/<unstructured/structured>

cilium encrypt/cgroups commands don't output json, moved them to md files.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>
